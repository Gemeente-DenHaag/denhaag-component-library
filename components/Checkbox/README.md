# Checkbox

![npm (scoped)](https://img.shields.io/npm/v/@gemeente-denhaag/checkbox?logo=npm&style=flat-square)
![GitHub Workflow Status (branch)](https://img.shields.io/github/workflow/status/nl-design-system/denhaag/Build%20and%20deploy%20Storybook%20to%20Azure%20Web%20App/main?logo=github&style=flat-square)

Checkboxes are usually used to provide a way for users to make a range of selections (zero, one or more). They can also be used to tell users that they agree to specific terms.

## When to use

Checkboxes are used for multiple choices, not for mutually exclusive choices. Each checkbox works independently from other checkboxes in the list, therefore checking an additional box does not affect any other selections.

### Forms

Can be used in forms on a full page or in modals.

### Terms and conditions

Turning the checkbox input on or off can indicate whether you agree to the terms.

## Alternatives and related components

Use radio buttons to display a list of options where users can only make one choice

[Navigate to Radio](../?path=/docs/components-input-radio--default)

## Anatomy

The checkbox consists of:

1. Label (optional): indicates what should be selected below.
2. Checkbox input: a checkbox input that indicates the correct status. By default it is not active.
3. Checkbox label: describes the information you can select or deselect.
4. Helper text (optional): displays more information about the checkboxes

## (Interactive) states

The checkbox contains the states normal, hover, disabled, error and focus.

## Design properties

### Typography

- Label: TheSans/md/700
- Checkbox label: TheSans/md/400
- Helper text: TheSans/s/400

### Colors

- Label: text color Blue/5
- Checkbox input inactive: border color Grey/4, background-color white
- Checkbox input active: border color Blue/3, background-color Blue/3, checkmark icon white
- Checkbox label: text color Grey/4
- Helper text: text color Grey/4

### Interactive states

- Disabled: container border color Grey/2, checkbox label text color Grey/2
- Error: container border color Red/3, checkbox label text color Grey/4
- Focus: container border color Ocher/5

### Structure

- Checkbox: height and width 20px, border 1px
- Checkbox focus: border 2px, dash 5,5
- Label: margin-bottom 8px
- Checkbox label: padding-left 8px
- Helper text: margin-left 16px, margin-top 4px

## Accessibility

Screen readers automatically report the status of the check box.

- [technical accessibility requirements]

### Labeling

[technical requirements, such as: label and labelHidden]

### External links

[technical requirements]

### Keyboard support

- Move focus to each checkbox using the tab key (or shift + tab when tabbing backwards)
- Activate or deactivate checkboxes with the space key

## Content guidelines

### Checkbox labels

Checkbox labels must:

- Start with a capital letter
- Not end in punctuation if it’s a single sentence, word, or a fragment
- In case the checkbox asks to agree to the terms and conditions, use the first person

## Best practices

### Do's

Checkboxes must:

- Work independently of each other: selecting a checkbox may not change the selection status of another checkbox in the list.
- Be positively boxed: for example “Enable notifications” instead of “Disable notifications”.
- Always have a label handy when you use it to turn a setting on or off.
- Use label tags as click targets.
- Be arranged in a logical order, whether alphabetical, numeric, time-based or some other clear system.
- Link to more information or have a subtitle for more explanation. Don't rely on tooltips to explain a checkbox.
- If they are too long, run over the second line. This is preferable than truncation.
- If they are too long, they will be aligned left
- If they are too long, let the text flow under the checkbox so that the control and label are aligned at the top.

### Don'ts

Checkboxes must:

- Not cut with an omit task if the label is too long and run over 2 lines.

### Buttons versus links

Buttons are mainly used for actions such as "Add", "Close", "Cancel" or "Save". Regular buttons similar to links are used for less important or less frequently used actions, such as "View details".

Links are mainly used for navigation and usually appear in or immediately after a sentence.

The HTML displayed for the Button and Link components has meaning. Using these components intentionally and consistently results in:

- a more inclusive experience for assistive technology users
- a more coherent visual experience for sighted users
- products that are easier to maintain to scale

## References

https://www.nngroup.com/articles/checkboxes-vs-radio-buttons/ </br>
https://uxplanet.org/checkbox-and-toggle-in-forms-f0de6086ac41 </br>
