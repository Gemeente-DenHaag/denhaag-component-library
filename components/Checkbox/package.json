{
  "name": "@gemeente-denhaag/checkbox",
  "version": "0.2.3-alpha.160",
  "description": "A Checkbox component",
  "bugs": "https://github.com/nl-design-system/denhaag/issues",
  "repository": {
    "type": "git",
    "url": "https://github.com/nl-design-system/denhaag.git",
    "directory": "components/Checkbox"
  },
  "license": "EUPL-1.2",
  "author": "Municipality of The Hague",
  "exports": "./dist/index.js",
  "main": "dist/index.js",
  "module": "dist/index.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "rollup -c ../../rollup.config.js",
    "clean": "yarn rimraf dist tsconfig.tsbuildinfo"
  },
  "dependencies": {
    "@gemeente-denhaag/icons": "0.2.3-alpha.160"
  },
  "devDependencies": {
    "@gemeente-denhaag/form-field": "0.1.1-alpha.36",
    "@gemeente-denhaag/formcontrollabel": "0.2.3-alpha.160"
  },
  "peerDependencies": {
    "react": "^17.0.0"
  },
  "gitHead": "dcf72a9b79266c1ebede35aff4a02dd9121a980f"
}
