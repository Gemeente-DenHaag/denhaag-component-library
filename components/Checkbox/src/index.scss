@import "~@utrecht/components/dist/checkbox/css";

.utrecht-checkbox--custom {
  border-radius: var(--utrecht-checkbox-border-radius);
}

.utrecht-checkbox--disabled {
  background-color: var(--utrecht-checkbox-disabled-background-color);
}
