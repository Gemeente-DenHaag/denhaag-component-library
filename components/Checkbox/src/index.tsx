import React, { InputHTMLAttributes } from 'react';
import clsx from 'clsx';
import { CheckedIcon } from '@gemeente-denhaag/icons';

import './index.scss';

export interface CheckboxProps extends InputHTMLAttributes<HTMLInputElement> {
  /**
   * Displays the checkbox red.
   */
  error?: boolean;
  /**
   * Disables Checkbox
   */
  disabled?: boolean;
  /**
   * Callback fired when the state is changed.
   */
  onChange?: (event: React.ChangeEvent<unknown>) => void;
  /**
   * Overwrite the Icon shown when in the checked state.
   */
  checkedIcon?: React.ReactNode;
}

/**
 * Checkboxes allow the user to select one or more items from a set.
 */
export const Checkbox: React.FC<CheckboxProps> = ({
  className,
  error = false,
  checked = false,
  checkedIcon = <CheckedIcon />,
  onChange,
  ...props
}: CheckboxProps) => {
  const [isChecked, setChecked] = React.useState(checked);
  const classes = clsx('denhaag-checkbox', { 'denhaag-checkbox--error': error }, className);

  return (
    <span className={classes}>
      <input
        type="checkbox"
        className="denhaag-checkbox__input"
        checked={isChecked}
        onChange={(event) => {
          setChecked(!isChecked);
          if (onChange) {
            onChange(event);
          }
        }}
        {...props}
      />
      <span className="denhaag-checkbox__icon">{checkedIcon}</span>
    </span>
  );
};

export default Checkbox;
