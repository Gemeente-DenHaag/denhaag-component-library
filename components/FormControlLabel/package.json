{
  "name": "@gemeente-denhaag/formcontrollabel",
  "version": "0.2.3-alpha.368",
  "description": "A FormControlLabel component",
  "bugs": "https://github.com/nl-design-system/denhaag/issues",
  "repository": {
    "type": "git",
    "url": "https://github.com/nl-design-system/denhaag.git",
    "directory": "components/FormControlLabel"
  },
  "license": "EUPL-1.2",
  "author": "Municipality of The Hague",
  "exports": {
    ".": {
      "types": "./dist/cjs/index.d.ts",
      "import": "./dist/mjs/index.js",
      "default": "./dist/cjs/index.js"
    },
    "./index.css": "./dist/index.css"
  },
  "main": "dist/cjs/index.js",
  "module": "dist/mjs/index.js",
  "types": "dist/cjs/index.d.ts",
  "files": [
    "dist"
  ],
  "devDependencies": {
    "@utrecht/components": "1.0.1-alpha.2"
  },
  "scripts": {
    "build": "rollup -c ../../rollup.config.mjs",
    "clean": "rimraf dist tsconfig.tsbuildinfo",
    "watch:components": "chokidar --silent --follow-symlinks \"src/**/*\" -c \"pnpm run build\""
  },
  "peerDependencies": {
    "react": "^18.0.0"
  },
  "gitHead": "dcf72a9b79266c1ebede35aff4a02dd9121a980f"
}
