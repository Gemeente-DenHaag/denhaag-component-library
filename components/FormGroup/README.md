# FormGroup

![npm (scoped)](https://img.shields.io/npm/v/@gemeente-denhaag/formgroup?logo=npm&style=flat-square)
![GitHub Workflow Status (branch)](https://img.shields.io/github/workflow/status/Gemeente-Denhaag/denhaag-component-library/Build%20and%20deploy%20Storybook%20to%20Azure%20Web%20App/master?logo=github&style=flat-square)

This component is used together with FormControlLabel and CheckBox or Radio.
It ensures consistent spacing between Control elements (CheckBox or Radio) and can place Heading and helpertext labels.

For an example of how this component is used, see [Checkbox](../?path=/docs/components-input-checkbox--default)
