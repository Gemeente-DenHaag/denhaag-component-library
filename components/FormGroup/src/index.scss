.denhaag-form-group {
  /* Overwrite utrecht design tokens */
  --utrecht-heading-5-color: var(--denhaag-form-group-label-color);
  --utrecht-paragraph-font-size: var(--denhaag-form-group-helper-text-size);
  --utrecht-paragraph-color: var(--denhaag-form-group-helper-text-color);

  /* To ensure gap between helpertext and checkboxes */
  --utrecht-paragraph-margin-block-start: var(--denhaag-form-group-helper-text-margin-block-start);

  /* To ensure gap between checkboxes */
  --denhaag-form-control-label-margin-block-end: 0;
  --denhaag-form-control-label-margin-inline-start: 0;

  display: flex;
  flex-direction: column;
  margin-inline-start: var(--denhaag-form-group-margin-inline-start);
  margin-inline-end: var(--denhaag-form-group-margin-inline-end);
  margin-block-start: var(--denhaag-form-group-margin-block-start);
  margin-block-end: var(--denhaag-form-group-margin-block-end);
}

.denhaag-form-group--error {
  --utrecht-paragraph-color: var(--denhaag-form-group-helper-text-error-color);
}
