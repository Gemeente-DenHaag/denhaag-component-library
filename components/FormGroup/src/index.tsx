import React from 'react';
import clsx from 'clsx';
import { Heading5, Paragraph } from '@gemeente-denhaag/typography';

import './index.scss';

export interface FormGroupProps {
  /**
   * The contents of the component.
   */
  children?: React.ReactNode;
  /**
   * The formgroup label shown at the top
   */
  label?: string;

  /**
   * The helpertext shown at the bottom.
   */
  helperText?: string;

  /**
   * When true, turns the helperText red.
   */
  error?: boolean;
}

/**
 *  FormGroup wraps controls such as Checkbox and Switch.
 *  For the Radio, you should be using the RadioGroup component instead of this one.
 */
export const FormGroup: React.FC<FormGroupProps> = (props: FormGroupProps) => {
  const rootStyles = clsx('denhaag-form-group', {
    'denhaag-form-group--error': props.error,
  });

  return (
    <div className={rootStyles}>
      {props.label && <Heading5>{props.label}</Heading5>}
      {props.children}
      {props.helperText && <Paragraph>{props.helperText}</Paragraph>}
    </div>
  );
};

export default FormGroup;
