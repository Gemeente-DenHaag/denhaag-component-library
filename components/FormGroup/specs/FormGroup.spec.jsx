import * as React from 'react';
import FormGroup from '../dist';
import Checkbox from '../../Checkbox';
import FormControlLabel from '../../FormControlLabel';

describe(`FormGroup tests`, () => {
  it('a11y and snapshots', () => {
    cy.snapshots(
      <FormGroup label="Label" helperText="Helper/Error text">
        <FormControlLabel label="Checkbox" control={<Checkbox error checked></Checkbox>}></FormControlLabel>
        <FormControlLabel label="Checkbox" control={<Checkbox disabled></Checkbox>}></FormControlLabel>
        <FormControlLabel label="Checkbox" control={<Checkbox checked></Checkbox>}></FormControlLabel>
        <FormControlLabel label="Checkbox" control={<Checkbox></Checkbox>}></FormControlLabel>
      </FormGroup>,
      {
        error: true,
      },
    );
  });
});
