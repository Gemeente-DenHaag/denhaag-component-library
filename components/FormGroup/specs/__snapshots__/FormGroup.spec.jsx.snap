exports[`FormGroup tests > a11y and snapshots #0`] = `
<div id="wrapper" style="display: inline-block;">
  <div style="padding: 10px;">
    <div class="denhaag-form-group">
      <h5 class="utrecht-heading-5 utrecht-heading-5--distanced">Label</h5>
      <label class="MuiFormControlLabel-root denhaag-form-control-label"
        ><span
          class="denhaag-checkbox denhaag-checkbox--error denhaag-checkbox--checked"
          ><input type="checkbox" class="denhaag-checkbox__input" checked=""/>
          <div class="denhaag-checkbox__icon">
            <svg
              xmlns="http://www.w3.org/2000/svg"
              width="1em"
              height="1em"
              viewBox="0 0 24 24"
              class="denhaag-icon"
              focusable="false"
              aria-hidden="true"
              shape-rendering="auto"
            >
              <rect
                x="-9.5"
                y="-9.5"
                rx="0.5"
                ry="0.5"
                width="19"
                height="19"
                stroke="currentColor"
                fill="currentColor"
                transform="translate(12 12)"
              ></rect>
              <path
                d="M17.663 8l-7.333 7.333L6.997 12"
                stroke-linecap="round"
                stroke="#fff"
                stroke-width="2"
                stroke-linejoin="round"
                fill="none"
              ></path>
            </svg></div></span
        ><span
          class="MuiTypography-root MuiFormControlLabel-label denhaag-form-control-label__label MuiTypography-body1"
          >Checkbox</span
        ></label
      ><label
        class="MuiFormControlLabel-root denhaag-form-control-label Mui-disabled denhaag-form-control-label--disabled"
        ><span class="denhaag-checkbox denhaag-checkbox--disabled"
          ><input disabled="" type="checkbox" class="denhaag-checkbox__input"/>
          <div class="denhaag-checkbox__icon">
            <svg
              xmlns="http://www.w3.org/2000/svg"
              width="1em"
              height="1em"
              viewBox="0 0 24 24"
              class="denhaag-icon"
              focusable="false"
              aria-hidden="true"
              shape-rendering="auto"
            >
              <rect
                x="-9.5"
                y="-9.5"
                rx="0.5"
                ry="0.5"
                width="19"
                height="19"
                stroke="currentColor"
                fill="#fff"
                transform="translate(12 12)"
              ></rect>
            </svg></div></span
        ><span
          class="MuiTypography-root MuiFormControlLabel-label denhaag-form-control-label__label Mui-disabled denhaag-form-control-label--disabled MuiTypography-body1"
          >Checkbox</span
        ></label
      ><label class="MuiFormControlLabel-root denhaag-form-control-label"
        ><span class="denhaag-checkbox denhaag-checkbox--checked"
          ><input type="checkbox" class="denhaag-checkbox__input" checked=""/>
          <div class="denhaag-checkbox__icon">
            <svg
              xmlns="http://www.w3.org/2000/svg"
              width="1em"
              height="1em"
              viewBox="0 0 24 24"
              class="denhaag-icon"
              focusable="false"
              aria-hidden="true"
              shape-rendering="auto"
            >
              <rect
                x="-9.5"
                y="-9.5"
                rx="0.5"
                ry="0.5"
                width="19"
                height="19"
                stroke="currentColor"
                fill="currentColor"
                transform="translate(12 12)"
              ></rect>
              <path
                d="M17.663 8l-7.333 7.333L6.997 12"
                stroke-linecap="round"
                stroke="#fff"
                stroke-width="2"
                stroke-linejoin="round"
                fill="none"
              ></path>
            </svg></div></span
        ><span
          class="MuiTypography-root MuiFormControlLabel-label denhaag-form-control-label__label MuiTypography-body1"
          >Checkbox</span
        ></label
      ><label class="MuiFormControlLabel-root denhaag-form-control-label"
        ><span class="denhaag-checkbox"
          ><input type="checkbox" class="denhaag-checkbox__input"/>
          <div class="denhaag-checkbox__icon">
            <svg
              xmlns="http://www.w3.org/2000/svg"
              width="1em"
              height="1em"
              viewBox="0 0 24 24"
              class="denhaag-icon"
              focusable="false"
              aria-hidden="true"
              shape-rendering="auto"
            >
              <rect
                x="-9.5"
                y="-9.5"
                rx="0.5"
                ry="0.5"
                width="19"
                height="19"
                stroke="currentColor"
                fill="#fff"
                transform="translate(12 12)"
              ></rect>
            </svg></div></span
        ><span
          class="MuiTypography-root MuiFormControlLabel-label denhaag-form-control-label__label MuiTypography-body1"
          >Checkbox</span
        ></label
      >
      <p class="utrecht-paragraph utrecht-paragraph--distanced">
        Helper/Error text
      </p>
    </div>
  </div>
  <div style="padding: 10px;">
    <div class="denhaag-form-group denhaag-form-group--error">
      <h5 class="utrecht-heading-5 utrecht-heading-5--distanced">Label</h5>
      <label class="MuiFormControlLabel-root denhaag-form-control-label"
        ><span
          class="denhaag-checkbox denhaag-checkbox--error denhaag-checkbox--checked"
          ><input type="checkbox" class="denhaag-checkbox__input" checked=""/>
          <div class="denhaag-checkbox__icon">
            <svg
              xmlns="http://www.w3.org/2000/svg"
              width="1em"
              height="1em"
              viewBox="0 0 24 24"
              class="denhaag-icon"
              focusable="false"
              aria-hidden="true"
              shape-rendering="auto"
            >
              <rect
                x="-9.5"
                y="-9.5"
                rx="0.5"
                ry="0.5"
                width="19"
                height="19"
                stroke="currentColor"
                fill="currentColor"
                transform="translate(12 12)"
              ></rect>
              <path
                d="M17.663 8l-7.333 7.333L6.997 12"
                stroke-linecap="round"
                stroke="#fff"
                stroke-width="2"
                stroke-linejoin="round"
                fill="none"
              ></path>
            </svg></div></span
        ><span
          class="MuiTypography-root MuiFormControlLabel-label denhaag-form-control-label__label MuiTypography-body1"
          >Checkbox</span
        ></label
      ><label
        class="MuiFormControlLabel-root denhaag-form-control-label Mui-disabled denhaag-form-control-label--disabled"
        ><span class="denhaag-checkbox denhaag-checkbox--disabled"
          ><input disabled="" type="checkbox" class="denhaag-checkbox__input"/>
          <div class="denhaag-checkbox__icon">
            <svg
              xmlns="http://www.w3.org/2000/svg"
              width="1em"
              height="1em"
              viewBox="0 0 24 24"
              class="denhaag-icon"
              focusable="false"
              aria-hidden="true"
              shape-rendering="auto"
            >
              <rect
                x="-9.5"
                y="-9.5"
                rx="0.5"
                ry="0.5"
                width="19"
                height="19"
                stroke="currentColor"
                fill="#fff"
                transform="translate(12 12)"
              ></rect>
            </svg></div></span
        ><span
          class="MuiTypography-root MuiFormControlLabel-label denhaag-form-control-label__label Mui-disabled denhaag-form-control-label--disabled MuiTypography-body1"
          >Checkbox</span
        ></label
      ><label class="MuiFormControlLabel-root denhaag-form-control-label"
        ><span class="denhaag-checkbox denhaag-checkbox--checked"
          ><input type="checkbox" class="denhaag-checkbox__input" checked=""/>
          <div class="denhaag-checkbox__icon">
            <svg
              xmlns="http://www.w3.org/2000/svg"
              width="1em"
              height="1em"
              viewBox="0 0 24 24"
              class="denhaag-icon"
              focusable="false"
              aria-hidden="true"
              shape-rendering="auto"
            >
              <rect
                x="-9.5"
                y="-9.5"
                rx="0.5"
                ry="0.5"
                width="19"
                height="19"
                stroke="currentColor"
                fill="currentColor"
                transform="translate(12 12)"
              ></rect>
              <path
                d="M17.663 8l-7.333 7.333L6.997 12"
                stroke-linecap="round"
                stroke="#fff"
                stroke-width="2"
                stroke-linejoin="round"
                fill="none"
              ></path>
            </svg></div></span
        ><span
          class="MuiTypography-root MuiFormControlLabel-label denhaag-form-control-label__label MuiTypography-body1"
          >Checkbox</span
        ></label
      ><label class="MuiFormControlLabel-root denhaag-form-control-label"
        ><span class="denhaag-checkbox"
          ><input type="checkbox" class="denhaag-checkbox__input"/>
          <div class="denhaag-checkbox__icon">
            <svg
              xmlns="http://www.w3.org/2000/svg"
              width="1em"
              height="1em"
              viewBox="0 0 24 24"
              class="denhaag-icon"
              focusable="false"
              aria-hidden="true"
              shape-rendering="auto"
            >
              <rect
                x="-9.5"
                y="-9.5"
                rx="0.5"
                ry="0.5"
                width="19"
                height="19"
                stroke="currentColor"
                fill="#fff"
                transform="translate(12 12)"
              ></rect>
            </svg></div></span
        ><span
          class="MuiTypography-root MuiFormControlLabel-label denhaag-form-control-label__label MuiTypography-body1"
          >Checkbox</span
        ></label
      >
      <p class="utrecht-paragraph utrecht-paragraph--distanced">
        Helper/Error text
      </p>
    </div>
  </div>
</div>
`;
